realm/realm-library/src/main/java/io/realm/BaseRealm.java

```diff
@@ -271,13 +271,13 @@ public void refresh() {
         if (isInTransaction()) {
             throw new IllegalStateException(BaseRealm.CANNOT_REFRESH_INSIDE_OF_TRANSACTION_MESSAGE);
         }
-        sharedGroupManager.advanceRead();
-        if (handlerController != null) {
-            handlerController.notifyAllListeners();
-            // if we have empty async RealmObject then rerun
-            if (handlerController.threadContainsAsyncEmptyRealmObject()) {
-                handlerController.updateAsyncEmptyRealmObject();
-            }
+        if (!handlerController.isAutoRefreshEnabled()) {
+            // non Looper Thread, just advance the Realm
+            // registering listeners is not allowed, hence nothing to notify
+            sharedGroupManager.advanceRead();
+            handlerController.refreshTableViews();
+        } else {
+            handlerController.notifyCurrentThreadRealmChanged();
```

Comment:
This is slightly different. This abstraction is there so it is only decided in one place if we trigger advance_read/notifications straight away or use the looper.